windows-gaming-driver/src/driver/clipboard.rs

```diff
@@ -0,0 +1,213 @@
+                screen.root_visual(),
+                &[(
+                    xcb::CW_EVENT_MASK,
+                    xcb::EVENT_MASK_STRUCTURE_NOTIFY | xcb::EVENT_MASK_PROPERTY_CHANGE
+                )]
+            );
+            connection.flush();
+        }
+
+        let clipboard = get_atom(&connection, "CLIPBOARD")?;
+        let property = get_atom(&connection, "THIS_CLIPBOARD_OUT")?;
+        let targets = get_atom(&connection, "TARGETS")?;
+        let utf8_string = get_atom(&connection, "UTF8_STRING")?;
+        // let incr = get_atom(&connection, "INCR")?; // not (yet?) implemented
+
+        Ok(X11Clipboard { connection, window, atoms: Atoms { clipboard, targets, utf8_string, property } })
+    }
+
+    pub fn grab_clipboard(&self) {
+        xcb::set_selection_owner(&self.connection, self.window, self.atoms.clipboard, xcb::CURRENT_TIME);
+        self.connection.flush();
+    }
+
+    pub fn read_clipboard(&self) {
+        xcb::convert_selection(&self.connection, self.window, self.atoms.clipboard, self.atoms.utf8_string,
+                               self.atoms.property, xcb::CURRENT_TIME);
+        self.connection.flush();
+    }
+
+    pub fn run<'a>(&'a self, handle: &Handle) -> ClipboardContext<'a> {
+        let (lost_send, lost_recv) = mpsc::unbounded();
+        let (query_send, query_recv) = mpsc::unbounded();
+        let (data_send, data_recv) = mpsc::unbounded();
+        let (resp_send, resp_recv) = mpsc::unbounded();
+
+        let listener = Box::new(XcbEvents::new(&self.connection, handle).for_each(move |event| {
+            match event.response_type() & !0x80 {
+                SELECTION_REQUEST => {
+                    let event: &SelectionRequestEvent = unsafe { xcb::cast_event(&event) };
+                    query_send.send(ClipboardRequestEvent {
+                        time: event.time(),
+                        requestor: event.requestor(),
+                        selection: event.selection(),
+                        target: event.target(),
+                        property: event.property(),
+                    }).unwrap();
+                }
+                SELECTION_CLEAR => {
+                    lost_send.send(()).unwrap();
+                }
+                SELECTION_NOTIFY => {
+                    let event: &SelectionNotifyEvent = unsafe { xcb::cast_event(&event) };
+                    let reply = xcb::get_property(
+                        &self.connection, false, self.window,
+                        event.property(), self.atoms.utf8_string, 0, ::std::u32::MAX // FIXME reasonable buffer size
+                    ).get_reply().unwrap(); // FIXME
+                    assert!(reply.type_() == self.atoms.utf8_string);
+                    data_send.send(reply.value().to_vec()).unwrap();
+                }
+                _ => unimplemented!(),
+            }
+
+            Ok(())
+        }).join(resp_recv.for_each(move |response: ClipboardRequestResponse| {
```

Comment:
Please create this future (and the above) as variable and join 2 variables with `future1.join(future2)`.